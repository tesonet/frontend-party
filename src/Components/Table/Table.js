import React from "react";

export const Table = ({ headings, items }) => (
  <div className="table">
    <div className="table__header">
      <div className="table__row">
        {headings &&
          headings.map((heading, index) => (
            <div className="table__cell" key={`${heading}-${index}`}>
              {heading}
            </div>
          ))}
      </div>
    </div>
    <div className="table__body">
      {items &&
        items.map((item, index) => (
          <div className="table__row" key={`${item.name}-${index}`}>
            <div className="table__cell">{item.name}</div>
            <div className="table__cell">{item.distance} km</div>
          </div>
        ))}
    </div>
  </div>
);
