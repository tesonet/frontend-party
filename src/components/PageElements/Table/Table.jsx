import React, { useState, useEffect } from 'react';
import { arrayOf, func, object } from 'prop-types';

import { Box } from 'components/Core';

import { Header } from './Header';
import { Row } from './Row';

const propTypes = {
  headerItems: arrayOf(object),
  items: arrayOf(object),
  sortingFunction: func
};

const defaultProps = {
  headerItems: [],
  items: []
};

const Table = ({ headerItems, items, sortingFunction }) => {
  const [renderedItems, setRenderedItems] = useState(items);
  const [sortBy, setSortBy] = useState();

  useEffect(() => {
    setRenderedItems(items);
  }, [items]);

  useEffect(() => {
    if (sortBy && sortingFunction) {
      setRenderedItems(sortingFunction(sortBy, items));
    }
  }, [sortBy]);

  return (
    <Box>
      <Header
        items={headerItems}
        onSort={props => {
          setSortBy(props);
        }}
      />
      {renderedItems.map(item => (
        <Row key={item.key} {...item} />
      ))}
    </Box>
  );
};

Table.defaultProps = defaultProps;
Table.propTypes = propTypes;

export default Table;
