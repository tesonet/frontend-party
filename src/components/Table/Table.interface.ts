export interface IAPIResponse {
  name: string;
  distance: number;
}
