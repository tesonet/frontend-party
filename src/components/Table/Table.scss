.table-header {
    background: #f5f5f5 !important;
    border-bottom: 1px !important;
}

.table-header-sort {
    cursor: pointer;
    font-weight: 200 !important;
    color: #a8a8a8;
}

.table-body {
    color: #6a6a6a;
}