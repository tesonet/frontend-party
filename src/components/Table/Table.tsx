import React, { useState, useEffect } from "react";
import styled from "styled-components";
import { Server } from "../../store/servers/actions";

interface TableHeaderColumn {
  key: string;
  display: string;
}

interface TableHeaderProps {
  columns: TableHeaderColumn[];
  onHeaderClick: Function;
}

interface TableProps {
  data: Server[];
  columns: TableHeaderColumn[];
}

const TableContainer = styled.div`
  display: flex;
  flex-direction: column;
  min-height: 100%;
  color: #999999;
`;

const TableHeaderContainer = styled.div`
  background-color: #f5f5f5;
  display: flex;
  flex-direction: row;
  justify-content: space-between;
`;

const Column = styled.div`
  padding: 1rem;
  cursor: default;
`;

const Row = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  border-bottom: 1px solid #e6e6e6;
`;

const TableHeader: React.FC<TableHeaderProps> = ({
  columns,
  onHeaderClick
}) => (
  <TableHeaderContainer>
    {columns.map(column => (
      <Column key={column.key} onClick={() => onHeaderClick(column.key)}>
        {column.display}
      </Column>
    ))}
  </TableHeaderContainer>
);

const TableBody = styled.div`
  flex: 1 1 auto;
  overflow-y: auto;
  height: 100px;
`;

enum SortDirections {
  "NONE" = null,
  "ASC" = 1,
  "DESC" = -1
}
type SortKeys = null | "name" | "distance";

export const Table: React.FC<TableProps> = ({ data, columns }) => {
  const [sortedData, setSortedData] = useState<Server[]>([]);
  const [sortKey, setSortKey] = useState<SortKeys>(null);
  const [sortDirection, setSortDirection] = useState<SortDirections>(null);

  useEffect(() => {
    const sort = (): Server[] => {
      if (sortDirection !== null && sortKey !== null) {
        return JSON.parse(JSON.stringify(data)).sort((a: Server, b: Server) => {
          if (a[sortKey] < b[sortKey]) {
            return -1 * sortDirection;
          }
          if (a[sortKey] > b[sortKey]) {
            return 1 * sortDirection;
          }
          return 0;
        });
      } else {
        return data;
      }
    };
    setSortedData(sort());
  }, [data, sortKey, sortDirection]);

  const handleHeaderClick = (sortKey: SortKeys): void => {
    setSortKey(sortKey);
    setSortDirection((direction: SortDirections) =>
      direction !== null ? direction * -1 : SortDirections.ASC
    );
  };

  return (
    <TableContainer>
      <TableHeader columns={columns} onHeaderClick={handleHeaderClick} />
      <TableBody>
        {sortedData.map(row => (
          <Row key={`${row.name}-${row.distance}`}>
            <Column>{row.name}</Column>
            <Column>{row.distance}</Column>
          </Row>
        ))}
      </TableBody>
    </TableContainer>
  );
};
