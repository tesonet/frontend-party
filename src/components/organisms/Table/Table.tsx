import React from 'react';
import styles from './Table.module.scss';

export type TAlignAttr = 'left' | 'center' | 'right' | 'justify' | 'char';

interface IColumnProps {
	key: string;
	align?: TAlignAttr;
	title: string;
	dataIndex: string;
}

export interface ITableProps {
	columns: Array<IColumnProps>
	data: Array<Record<string, any> & { key: string }>;
	tableStyles?: string;
}

export function Table({columns, data, tableStyles}: ITableProps) {
	return (
		<table className={[styles['table'], tableStyles].filter(Boolean).join(' ')}>
			<thead className={styles['table__thead']}>
				<tr className={styles['table__row']}>
				{
					columns.map(({key, align, title}) => (
						<th
							key={key}
							align={align}
							className={styles['table__header']}
						>
							{title}
						</th>
					))
				}
				</tr>
			</thead>
			<tbody className={styles['table__tbody']}>
			{
				data.map(row => (
					<tr key={row.key} className={styles['table__row']}>
						{
							columns.map(({key, align, dataIndex}) => (
								<td
									key={key}
									align={align}
									className={styles['table__cell']}
								>
									{row[dataIndex]}
								</td>
							))
						}
					</tr>
				))
			}
			</tbody>
		</table>
	);
}