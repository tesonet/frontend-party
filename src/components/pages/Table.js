import React from 'react';


const Table = () => (
    <div>
        <h1>Table</h1>
    </div>
)

export default Table;