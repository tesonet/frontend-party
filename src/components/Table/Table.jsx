
import React, { useEffect, useState } from 'react';
import { PropTypes } from 'prop-types';
import './Table.scss';

const Table = ({ servers }) => {
    const [tableRows, setTableRows] = useState([]);
    const [activeSort, setActiveSort] = useState({});

    const columns = [
        {
            header: 'SERVER',
            property: 'name'
        },
        {
            header: 'DISTANCE',
            property: 'distance'
        }
    ];

    const sort = (items, property, reverse) => {
        const sorted = items.sort((a, b) => {
            if (a[property] > b[property]) {
                return 1;
            }
            if (b[property] > a[property]) {
                return -1;
            }
            return 0;
        });

        return reverse ? sorted.reverse() : sorted;
    }

    const getTableRows = (items) => {
        const rows = items.map((item) => {
            return (
                <tr key={item.name + item.distance}>
                    <td>{item.name}</td>
                    <td className="text-right">{item.distance} km</td>
                </tr>
            );
        });

        return rows;
    }

    const handleSort = (property) => {
        const reverse = activeSort.property === property ? !activeSort.reverse : false;
        const sorted = sort(servers, property, reverse);
        const rows = getTableRows(sorted);

        setActiveSort({ property, reverse });
        setTableRows(rows);
    }

    useEffect(() => {
        handleSort(servers, 'name');
    }, [servers]);

    return (
        <table className="table">
            <thead>
                <tr>
                    {
                        columns.map((column) => {
                            return (
                                <th
                                    key={column.property}
                                    className={`table-header ${column.property === "distance" && "text-right"}`}
                                    onClick={() => handleSort(column.property)}>
                                    <span className="table-header-sort">
                                        {column.header}
                                        {column.property === activeSort.property &&
                                            <i className={activeSort.reverse ? "fas fa-caret-down" : "fas fa-caret-up"}></i>}
                                    </span>
                                </th>
                            );
                        })
                    }
                </tr>
            </thead>
            <tbody className="table-body">
                {tableRows}
            </tbody>
        </table>
    )
}

Table.defaultProps = {
    servers: [{ name: 'Default', distance: 0 }]
};

Table.propTypes = {
    servers: PropTypes.arrayOf(PropTypes.object)
}

export default Table;