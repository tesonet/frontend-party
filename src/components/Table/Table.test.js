import * as React from 'react';
import { configure, mount, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16'
import Table from "./Table";

configure({ adapter: new Adapter() });

describe('Table component', () => {
  const items = [{ name: 'Lithuania', distance: 10 }, { name: 'Latvia', distance: 20 }];

  it('should render component correctly', () => {
    const wrapper = shallow(<Table />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should receive 2 servers', () => {
    const wrapper = mount(<Table servers={items} />);
    expect(wrapper.props().servers).toHaveLength(items.length);
  });

  it('should show 2 column names', () => {
    const wrapper = shallow(<Table servers={items} />);
    expect(wrapper.find('th').length).toBe(2);
  });
});