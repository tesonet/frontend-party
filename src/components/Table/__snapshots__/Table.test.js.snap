// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table component should render component correctly 1`] = `
<table
  className="table"
>
  <thead>
    <tr>
      <th
        className="table-header false"
        key="name"
        onClick={[Function]}
      >
        <span
          className="table-header-sort"
        >
          SERVER
        </span>
      </th>
      <th
        className="table-header text-right"
        key="distance"
        onClick={[Function]}
      >
        <span
          className="table-header-sort"
        >
          DISTANCE
        </span>
      </th>
    </tr>
  </thead>
  <tbody
    className="table-body"
  />
</table>
`;
